lib/Core/Executor.cpp

```diff
@@ -3752,20 +3748,94 @@ void Executor::executeMarkPersistent(ExecutionState &state,
 
 void Executor::executeMarkPersistent(ExecutionState &state,
                                      const MemoryObject *mo) {
-  mo->isPersistent = true;
-  std::string name;
-  mo->getAllocInfo(name);
+  state.persistentObjects.insert(mo);
+
+  const ObjectState *os = state.addressSpace.findObject(mo);
+  assert(os && "Cannot mark unbound MemoryObject persistent");
+
+  // Create a symbolic bit array to track the state of each cache line.
+  // Initialize all cache lines to persisted (1).
+  unsigned nCacheLines = memory->getSizeInCacheLines(mo->size);
+  ref<ConstantExpr> Persisted = PersistentState::getPersistedExpr();
+  std::vector< ref<ConstantExpr> > InitialValues(nCacheLines, Persisted);
+  const std::string &baseName = mo->name;
+  const Array *cacheLines = arrayCache.CreateArray(
+      baseName + "_cacheLines", nCacheLines,
+      &InitialValues[0], &InitialValues[0] + InitialValues.size(),
+      Expr::Int32 /* domain */, Expr::Int8 /* range */);
+
+  // Construct a persistent ObjectState and bind it to the memory object.
+  state.addressSpace.bindObject(mo, new PersistentState(os, cacheLines));
+
+  /* std::string name; */
+  /* mo->getAllocInfo(name); */
   /* klee_message("Found non-volatile memory pointer: %s", name.c_str()); */
 }
 
+bool Executor::isPersistent(ExecutionState &state, const MemoryObject *mo) {
+  if (state.persistentObjects.count(mo)) {
+    const ObjectState *os = state.addressSpace.findObject(mo);
+    assert(dyn_cast<PersistentState>(os) != nullptr &&
+           "MemoryObject marked persistent but not bound to any PersistentState");
+    return true;
+  }
+  return false;
+}
+
 void Executor::executePersistentMemoryFlush(ExecutionState &state,
-                                            ref<ConstantExpr> address) {
-  uint64_t raw_addr = address->getZExtValue();
-  state.pmemState.flush(raw_addr);
+                                            ref<Expr> address) {
+  address = optimizer.optimizeExpr(address, true);
+
+  ObjectPair op;
+  bool success;
+  solver->setTimeout(coreSolverTimeout);
+  if (!state.addressSpace.resolveOne(state, solver, address, op, success)) {
+    address = toConstant(state, address, "resolveOne failure");
+    success = state.addressSpace.resolveOne(cast<ConstantExpr>(address), op);
+  }
+  solver->setTimeout(time::Span());
+
+  if (success) {
+    const MemoryObject *mo = op.first;
+    const ObjectState *os = op.second;
+    if (isPersistent(state, mo)) {
+      ObjectState *wos = state.addressSpace.getWriteable(mo, os);
+      PersistentState *ps = dyn_cast<PersistentState>(wos);
+      ref<Expr> offset = mo->getOffsetExpr(address);
+      ps->persistCacheLineAtOffset(offset);
+    }
+  } else {
+    terminateStateEarly(state, "Cannot singly resolve address to memory object");
+  }
```

Comment:
Shouldn't we check if it's a persistent state before trying to solve the address? Saves solver time.

Reply:
Can't check if it's persistent without a MemoryObject. Can't get a MemoryObject without resolving the address to a MemoryObject.

Every load or store makes this type of solver call. See `executeMemoryOperation`.

Reply:
Oh, I see. Makes sense.